File: Samples/Localization/C++/Src/ViewModel.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// NoesisGUI - http://www.noesisengine.com
// Copyright (c) 2013 Noesis Technologies S.L. All Rights Reserved.
////////////////////////////////////////////////////////////////////////////////////////////////////


#include "ViewModel.h"

#include <NsCore/ReflectionImplement.h>
#include <NsGui/ResourceDictionary.h>
#include <NsGui/ObservableCollection.h>
#include <NsGui/IntegrationAPI.h>


using namespace Localization;
using namespace Noesis;


////////////////////////////////////////////////////////////////////////////////////////////////////
Language::Language(const char* name, const char* sourcePath, Noesis::FlowDirection flow):
    _name(name), _source(sourcePath), _flow(flow)
{
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const char* Language::GetName() const
{
    return _name.Str();
}

////////////////////////////////////////////////////////////////////////////////////////////////////
const Uri& Language::GetSource() const
{
    return _source;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
Noesis::FlowDirection Language::GetFlowDirection() const
{
    return _flow;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
ViewModel::ViewModel()
{
    _languages = *new ObservableCollection<Language>();
    _languages->Add(MakePtr<Language>("English", 
        "Language-en.xaml", FlowDirection_LeftToRight));
    _languages->Add(MakePtr<Language>("Fran""\xC3\xA7""ais", 
        "Language-fr.xaml", FlowDirection_LeftToRight));
    _languages->Add(MakePtr<Language>("\xE6\x97\xA5\xE6\x9C\xAC\xE8\xAA\x9E", 
        "Language-jp.xaml", FlowDirection_LeftToRight));
    _languages->Add(MakePtr<Language>(u8"عربي", 
        "Language-ar.xaml", FlowDirection_RightToLeft));

    SetSelectedLanguage(_languages->Get(0));

    SetSoundLevel(100.0f);
    SetMusicLevel(70.0f);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
ObservableCollection<Language>* ViewModel::GetLanguages() const
{
    return _languages;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ViewModel::SetSelectedLanguage(Language* language)
{
    if (_selectedLanguage != language)
    {
        _selectedLanguage = language;
        OnPropertyChanged("SelectedLanguage");
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
Language* ViewModel::GetSelectedLanguage() const
{
    return _selectedLanguage;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ViewModel::SetSoundLevel(float level)
{
    if (_soundLevel != level)
    {
        _soundLevel = level;
        OnPropertyChanged("SoundLevel");
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float ViewModel::GetSoundLevel() const
{
    return _soundLevel;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ViewModel::SetMusicLevel(float level)
{
    if (_musicLevel != level)
    {
        _musicLevel = level;
        OnPropertyChanged("MusicLevel");
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float ViewModel::GetMusicLevel() const
{
    return _musicLevel;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
NS_BEGIN_COLD_REGION

NS_IMPLEMENT_REFLECTION(ViewModel)
{
    NsProp("Languages", &ViewModel::GetLanguages);
    NsProp("SelectedLanguage", &ViewModel::GetSelectedLanguage, &ViewModel::SetSelectedLanguage);
    NsProp("SoundLevel", &ViewModel::GetSoundLevel, &ViewModel::SetSoundLevel);
    NsProp("MusicLevel", &ViewModel::GetMusicLevel, &ViewModel::SetMusicLevel);
}

NS_IMPLEMENT_REFLECTION(Language)
{
    NsProp("Name", &Language::GetName);
    NsProp("Source", &Language::GetSource);
    NsProp("FlowDirection", &Language::GetFlowDirection);
}

NS_END_COLD_REGION

File: Samples/Localization/Unity/Assets/Samples/NoesisGUI/Localization/Blend~/ViewModel.cs
using System;
using System.Windows;
using System.Collections.ObjectModel;

namespace Localization
{
    public class Language
    {
        public string Name { get; set; }
        public Uri Source { get; set; }
        public FlowDirection FlowDirection { get; set; }
    }

    public class ViewModel : NotifyPropertyChangedBase
    {
        public ViewModel()
        {
            Languages = new ObservableCollection<Language>()
            {
                new Language
                {
                    Name = "English",
                    Source = new Uri("/Localization;component/Language-en.xaml", UriKind.Relative),
                    FlowDirection = FlowDirection.LeftToRight
                },

                new Language
                {
                    Name = "Français",
                    Source = new Uri("/Localization;component/Language-fr.xaml", UriKind.Relative),
                    FlowDirection = FlowDirection.LeftToRight
                },

                new Language
                {
                    Name = "日本語",
                    Source = new Uri("/Localization;component/Language-jp.xaml", UriKind.Relative),
                    FlowDirection = FlowDirection.LeftToRight
                },

                new Language
                {
                    Name = "عربي",
                    Source = new Uri("/Localization;component/Language-ar.xaml", UriKind.Relative),
                    FlowDirection = FlowDirection.RightToLeft
                }
            };

            SelectedLanguage = Languages[0];

            SoundLevel = 100.0f;
            MusicLevel = 80.0f;
        }

        public ObservableCollection<Language> Languages { get; private set; }

        private Language _selectedLanguage;
        public Language SelectedLanguage
        {
            get { return _selectedLanguage; }
            set
            {
                if (_selectedLanguage != value)
                {
                    _selectedLanguage = value;
                    OnPropertyChanged("SelectedLanguage");
                }
            }
        }

        private float _soundLevel;
        public float SoundLevel
        {
            get { return _soundLevel; }
            set
            {
                if (_soundLevel != value)
                {
                    _soundLevel = value;
                    OnPropertyChanged("SoundLevel");
                }
            }
        }

        private float _musicLevel;
        public float MusicLevel
        {
            get { return _musicLevel; }
            set
            {
                if (_musicLevel != value)
                {
                    _musicLevel = value;
                    OnPropertyChanged("MusicLevel");
                }
            }
        }
    }
}
